### ISSUE TITLE
Introduce Invariant Dependencies

### ISSUE THREAD

**Issue by davidstrauss** [systemd/systemd#4263]
An [invariant dependency](https://en.wikipedia.org/wiki/Invariant_%28computer_science%29) would be defined by Boolean expressions that are each (a) always satisfied or (b) in the process of becoming satisfied.

Let's say we have `multi-user.target`, `php.service`, and `mysql.service`. `php.service` specifies `RequiresRunning=mysql.service` and `RunningRequiredBy=multi-user.target` (which is just the inverse of `RequiresRunning`).

We now have three Boolean conditions:
1. _Always satisfied:_ running("mysql.service") || !running("php.service")
2. _Always satisfied:_ running("php.service") || !reached("multi-user.target")
3. _Opportunistically satisfied:_ reached("multi-user.target")

PID 1 would always want the "opportunistic" condition satisfied, but it's willing to temporarily violate it to satisfy the "always" ones. The target(s) that are opportunistically satisfied would be specified with something like `systemd.unit=`, a similar kernel parameter at boot, or an assumed default, while others could be added, even ephemerally, during runtime.

If services get added to the graph during runtime, systemd must still preserve invariant conditions. Adding a new service that injects a dependency to another (say, a new `nfs.service` specifying `RunningRequiredBy=mysql.service`) would need to stop the newly dependent unit (`mysql.service`), start the new unit (`nfs.service`), and then start the dependent unit again to re-satisfy the "opportunistic" condition (probably a target that requires `mysql.service` to be running) while never violating any "always" condition.

With multiple services specifying `RunningRequiredBy=multi-user.target` (usually by being enabled), some may fail. We still try to get every unit as close to being satisfied as possible, which creates a result similar to the existing `WantedBy=`. The main difference is that `multi-user.target` would only be in a totally "green" state if it's 100% satisfied after exhausting all efforts to make it so.

In fact, there are two types of failed states:
- An "always" condition has been violated (which may be recoverable but it at least for some time a failure).
- An "opportunistic" condition becomes impossible to achieve.

Being in a failed state would not excuse PID 1 from trying to maximize the number of satisfied conditions, prioritizing the "always" ones. So, if `mysql.service` fails, it must stop `php.service` and mark `multi-user.target` as unreachable, which results in the "always" conditions being satisfied even though the "opportunistic" condition becomes impossible.

To respond to unreachable "opportunistic" conditions, we would also introduce `OnImpossible=` for targets, which would specify a unit that starts if the target being configured cannot be satisfied. This would typically be configured on targets to automate recovery (like rebooting or notifying administrators if critical targets can't be satisfied).

**Comment by poettering** [comment]
Yeah, I think adding the concept of invariant deps makes a lot of sense, but I think we should literally call the dep type `Invariant=` or so.

**Comment by eliasp** [comment]
Adding my use-case here:
I have two services:
1. A long-running `salt-minion`
2. A one-shot `salt-minion-cache-cleaner`, triggered weekly through a timer
The latter one needs to temporarily shut down the former one for a few seconds to prevent races/cache corruption, but after the cache-cleaner is finished, `salt-minion` should start again.

My initial attempt was to use `Conflicts=salt-minion.service` in `salt-minion-cache-cleaner.service` and `Restart=always` in `salt-minion.service`, but this didn't have the desired effect, as [an explicit systemd-triggered stop](https://www.freedesktop.org/software/systemd/man/systemd.service.html#Restart=) of an service is exempt from `Restart=always`, so the result was a dead `salt-minion`.

**Comment by smurfix** [comment]
Is anybody working on this?
